### PR TITLE
feat(tap-runner): extend tap configuration

### PR DESCRIPTION
**Pull request by danny12321**
Extend the ability to configure the tap-runner to support frameworks that use their own runtime and support tap output. This feature request came from #4266 to support the Ava test framework.

### CODE REVIEW

File: packages/tap-runner/src/tap-helper.ts [hunk 1]
```diff
@@ -58,3 +58,27 @@ export async function captureTapResult(tapProcess: ChildProcessWithoutNullStream
   }
   return tapResult;
 }
+
+export function parseArguments(args: string[], hookFile: string, testFile: string): string[] {
```
Comment (nicojs) [inline packages/tap-runner/src/tap-helper.ts:62]: It's not really parsing. How about `buildArguments`?

File: packages/tap-runner/src/tap-helper.ts [hunk 2]
```diff
@@ -58,3 +58,27 @@ export async function captureTapResult(tapProcess: ChildProcessWithoutNullStream
...
+    if (index >= 0) {
+      newArgs[index] = newArgs[index].replace(arg.key, arg.replacement);
+    } else {
+      if (arg.placeInFront) {
+        newArgs.unshift(...arg.addition);
+      } else {
+        newArgs.push(...arg.addition);
+      }
+    }
+  });
```
Comment (nicojs) [inline packages/tap-runner/src/tap-helper.ts:81]: This seems pretty complicated. How about: 

```suggestion
const hooksFilePlaceholder = '{{hooksFile}}';
const testFilePlaceholder = '{{testFile}}';
if(!args.some(arg => arg.includes(hooksFilePlaceholder )){
   args = ['-r', hooksFilePlaceholder , ...args];
}
if(!args.some(arg => arg.includes(testFilePlaceholder ){
  args = [...args, testFilePlaceholder];
}
return args.map(arg => arg.replaceAll(testFilePlaceholder, testFile).replaceAll(hooksFilePlaceholder, hooksFile);
```

File: packages/tap-runner/test/unit/tap-helper.spec.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+      // Assert
+      expect(result).to.eql(['-r', 'hook.js', 'custom-arg', 'test.js']);
+    });
+
+    it('should replace {{hookFile}} and {{testFile}} in the arguments', () => {
+      const result = parseArguments(['--file={{hookFile}}', '--test={{testFile}}'], 'hook.js', 'test.js');
+
+      // Assert
+      expect(result).to.eql(['--file=hook.js', '--test=test.js']);
```
Comment (nicojs) [inline packages/tap-runner/test/unit/tap-helper.spec.ts:18]:
```suggestion
      expect(result).deep.eql(['--file=hook.js', '--test=test.js']);
```

File: packages/tap-runner/test/unit/tap-helper.spec.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { expect } from 'chai';
+
+import { parseArguments } from '../../src/tap-helper.js';
+
+describe('tap-helper', () => {
```
Comment (nicojs) [inline packages/tap-runner/test/unit/tap-helper.spec.ts:5]: Great to see some more unit tests 🤩.

Maybe also add tests for multiple instances of the placeholder.

File: packages/tap-runner/test/integration/tap-test-runner.it.spec.ts [hunk 5]
```diff
@@ -208,4 +208,41 @@ describe('tap-runner integration', () => {
...
+        const run = await sut.dryRun(factory.dryRunOptions());
+
+        // Assert
+        assertions.expectCompleted(run);
+      }
+    });
+
+    function getAvaLocation() {
+      if (os.platform() === 'win32') {
+        return `${sandbox.tmpDir}/../../../node_modules/ava/entrypoints/cli.mjs`;
```
Comment (nicojs) [inline packages/tap-runner/test/integration/tap-test-runner.it.spec.ts:242]: This should actually also work for linux, right?

Comment (nicojs) [inline packages/tap-runner/test/integration/tap-test-runner.it.spec.ts:242]:
```suggestion
        return path.resolve(sandbox.tmpDir, '..', '..', '..', 'node_modules', 'ava', 'entrypoints', 'cli.mjs');
```

Maybe a small improvement?

File: packages/tap-runner/src/tap-helper.ts (hunk 1 above)
Comment (danny12321) [reply packages/tap-runner/src/tap-helper.ts:62]: My thoughts was that it parses the template variables, but it indeed also builds the arguments if it is incomplete. I don't know a better name for now than `buildArguments`, but it isn't describing all the functionality in my opinion.

File: packages/tap-runner/test/integration/tap-test-runner.it.spec.ts (hunk 5 above)
Comment (danny12321) [reply packages/tap-runner/test/integration/tap-test-runner.it.spec.ts]: Indeed. Will change this to only use the full path (like on windows).

File: packages/tap-runner/test/integration/tap-test-runner.it.spec.ts [hunk 6]
```diff
@@ -208,4 +208,33 @@ describe('tap-runner integration', () => {
...
+      });
+      sut = testInjector.injector.injectFunction(createTapTestRunnerFactory('__stryker2__'));
+      await sut.init();
+    });
+
+    it('should be able to run', async function () {
+      // Test fails if debugger is used, because ava doesn't support debugging and tap output simultaneously.
+      // See: https://github.com/avajs/ava/blob/568fe40c987dd6c593dfbcf4144d1d1627955d46/lib/cli.js#L307
+      const { default: inspector } = await import('node:inspector');
+
```
Comment (nicojs) [inline packages/tap-runner/test/integration/tap-test-runner.it.spec.ts:229]: This is not needed. What we should do instead is disable `"autoAttachChildProcesses"` in launch.json. See https://rcrdo.com/2018/08/07/favorite-new-visual-studio-code-features-of-2018/#_node-debugging

**Pull request merged**